qcore/gsf.py

```diff
@@ -0,0 +1,211 @@
+"""This module provides functions for working with and generating GSF files.
+
+Functions:
+- gridpoint_count_in_length:
+    Calculates the number of gridpoints that fit into a given length.
+- coordinate_meshgrid:
+    Creates a meshgrid of points in a bounded plane region.
+- write_fault_to_gsf_file:
+    Writes geometry data to a GSF file.
+- read_gsf:
+    Parses a GSF file into a pandas DataFrame.
+
+The GSF file format is used to define the grid of the source model for a fault.
+See https://wiki.canterbury.ac.nz/display/QuakeCore/File+Formats+Used+On+GM
+for details on the GSF format.
+"""
+
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+
+from qcore import coordinates
+
+
+def gridpoint_count_in_length(length: float, resolution: float) -> int:
+    """Calculate the number of gridpoints that fit into a given length.
+
+    Computes the number of gridpoints that fit into a given length, if each
+    gridpoint is roughly resolution metres apart, and if the gridpoints
+    includes the endpoints. If length = 10, and resolution = 5, then the
+    function returns 3.
+
+      5m    5m
+    +-----+-----+
+
+    Parameters
+    ----------
+    length : float
+        Length to distribute grid points along.
+    resolution : float
+        Resolution of the grid.
+
+    Returns
+    -------
+    int
+        The number of gridpoints that fit into length.
+    """
+    return int(np.round(length / resolution + 2))
+
+
+def coordinate_meshgrid(
+    origin: np.ndarray,
+    x_upper: np.ndarray,
+    y_bottom: np.ndarray,
+    resolution: float,
+) -> np.ndarray:
+    """Creates a meshgrid of points in a bounded plane region.
+
+    Given the bounds of a rectangular planar region, create a meshgrid of
+    (lat, lon, depth) coordinates spaced at close to resolution metres apart
+    in the strike and dip directions.
+
+    origin                       x_upper
+          ┌─────────────────────┐
+          │. . . . . . . . . . .│
+          │                     │
+          │. . . . . . . . . . .│
+          │                     │
+          │. . . . . . . . . . .│
+          │                     │
+          │. . . . . . . . . . .│
+          │                     │
+          │. . . . . . . . . . .│
+          │            ∧ ∧      │
+          └────────────┼─┼──────┘
+     y_bottom          └─┘
+                    resolution
+
+    Parameters
+    ----------
+    origin : np.ndarray
+        Coordinates of the origin point (lat, lon, depth).
+    x_upper : np.ndarray
+        Coordinates of the upper x boundary (lat, lon, depth).
+    y_bottom : np.ndarray
+        Coordinates of the bottom y boundary (lat, lon, depth).
+    resolution : float
+        Resolution of the meshgrid.
+
+    Returns
+    -------
+    np.ndarray
+        The meshgrid of the rectangular planar region.
+    """
+    origin = coordinates.wgs_depth_to_nztm(origin)
```

Comment:
I think it might be worth breaking this up into some blocks, and then adding some block comments. There is a fair few steps and not all of them are trivial imo.